sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs

```diff
@@ -0,0 +1,168 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+using System.Collections.Immutable;
+using System.Text.RegularExpressions;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class UriShouldNotBeHardcodedBase<TExpressionSyntax,
+        TLiteralExpressionSyntax, TLanguageKindEnum, TBinaryExpressionSyntax, TArgumentSyntax, TVariableDeclaratorSyntax>
+        : SonarDiagnosticAnalyzer
+        where TExpressionSyntax : SyntaxNode
+        where TLiteralExpressionSyntax : TExpressionSyntax
+        where TBinaryExpressionSyntax : TExpressionSyntax
+        where TArgumentSyntax : SyntaxNode
+        where TVariableDeclaratorSyntax : SyntaxNode
+        where TLanguageKindEnum : struct
+    {
+        protected const string DiagnosticId = "S1075";
+        protected const string MessageFormat = "{0}";
+        private const string AbsoluteUriMessage = "Refactor your code not to use hardcoded absolute paths or URIs.";
+        private const string PathDelimiterMessage = "Remove this hardcoded path-delimiter.";
+
+        // Simplified implementation of specification listed on
+        // https://en.wikipedia.org/wiki/Uniform_Resource_Identifier
+        private const string UriScheme = "^[a-zA-Z][a-zA-Z\\+\\.\\-]+://.+";
+        private const string AbsoluteDiskUri = @"^[A-Za-z]:(/|\\)";
+        private const string AbsoluteMappedDiskUri = @"^\\\\\w[ \w\.]*";
+        private const string AbsoluteUnixUri = @"^(~\\|~/|/)\w";
+
+        private static readonly Regex UriRegex =
+            new Regex($"{UriScheme}|{AbsoluteDiskUri}|{AbsoluteMappedDiskUri}|{AbsoluteUnixUri}",
+                RegexOptions.Compiled);
+
+        private static readonly Regex PathDelimiterRegex = new Regex(@"^(\\|/)$", RegexOptions.Compiled);

```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A static field in a generic type is not shared among instances of different close constructed types. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2743)